swift/ql/src/queries/Security/CWE-135/StringLengthConflation.ql

```diff
@@ -11,5 +11,78 @@
  */
 
 import swift
+import codeql.swift.dataflow.DataFlow
+import DataFlow::PathGraph
 
-select "TODO"
+class StringLengthConflationConfiguration extends DataFlow::Configuration {
+  StringLengthConflationConfiguration() { this = "StringLengthConflationConfiguration" }
+
+  override predicate isSource(DataFlow::Node node, string flowstate) {
+    // result of a call to to `String.count`
+    exists(MemberRefExpr member |
+      member.getBaseExpr().getType().toString() = "String" and // TODO: use of toString
+      member.getMember().toString() = "count" and // TODO: use of toString
+      node.asExpr() = member and
+      flowstate = "String"
+    )
+  }
+
+  override predicate isSink(DataFlow::Node node, string flowstate) {
+    // arguments to method calls...
+    exists(
+      string className, string methodName, string argName, ClassDecl c, AbstractFunctionDecl f,
+      CallExpr call, int arg
+    |
+      (
+        // `NSRange.init`
+        className = "NSRange" and
+        methodName = "init" and
+        argName = ["location", "length"]
+        or
+        // `NSString.character`
+        className = ["NSString", "NSMutableString"] and
+        methodName = "character" and
+        argName = "at"
+        or
+        // `NSString.character`
+        className = ["NSString", "NSMutableString"] and
+        methodName = "substring" and
+        argName = ["from", "to"]
+        or
+        // `NSMutableString.insert`
+        className = "NSMutableString" and
+        methodName = "insert" and
+        argName = "at"
+      ) and
+      c.toString() = className and // TODO: use of toString
+      c.getAMember() = f and // TODO: will this even work if its defined in a parent class?
```

Comment:
No, it won't. I don't think we've implemented anything with the inheritance tree yet.

Reply:
I'm now wondering if it matters.  Is something like the following a situation we have to be concerned for in Swift?
```
class NSStringBase : NSObject
{
    // some of the implementation of NSString here
}

class NSString : NSStringBase
{
    // rest of the implementation of NSString
}
```
... or can we safely assume the implementations are in the places the docs say they are?

Reply:
I think that's something time will have to tell 😂. For now I think it's fine to assume the docs are precise about it

Reply:
NSString is actually an Objective-C class, so it's probably something we need to worry about.

Reply:
That's true. Looking at the ClassDecl for NSString in an image I have of Swift it doesn't look like it has any members at all :( this might be because it's defined in Obj-C code, or it might be due to some other extractor weirdness. In any case, the query right now might not work as well on a real snapshot as it does on the tests

Reply:
Thanks, I'll keep that in mind when I test on real code, which is high on my priorities right now.